text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,271 @@
+
+If an error occurs again, we can assume its proximate cause is in userland and not in LWC.
+
+
+### `hydrateElement`
+
+In a typical LWC web application, a developer might mount their app using the following pattern:
+
+```javascript
+import { createElement } from 'lwc';
+import MyApp from './my-app';
+
+document
+  .querySelector("#root")
+  .appendChild(createElement('my-app', { is: MyApp }));
+```
+
+After the proposed change, if a developer wants to take advantage of rehydration, a developer would instead do:
+
+```javascript
+import { hydrateElement } from 'lwc';
+import MyApp from './my-app';
+
+hydrateElement(document.querySelector('#root'), MyApp, { some: props }); 
+```
+
+`hydrateElement` accepts three arguments:
+
+1. The already-existing DOM node where the root component should be attached.
+2. The LWC class intended for use as the root component.
+3. Any props that should be supplied to that root component as part of the initial client-side render.
+
+### Constructable stylesheets
+
+The introduction of constructable stylesheets to LWC presents a small wrinkle. Since support for this browser API is not yet ubiquitous, we cannot know during SSR whether it will be supported on the client. For that reason, constructable stylesheets are disabled during SSR.
+
+From there, we have two options if support for constructable stylesheets is detected during rehydration:
+
+1. Swap in constructable stylesheets for vanilla stylesheets.
+2. Permanenly disable constructable stylesheets for the lifetime of the `document`.
+
+It is believed that swapping in constructable stylesheets could obviate the performance wins that constructable stylesheets are intended to bring. Furthermore, any theoretical benefit is outweighed by the implementation complexity of swapping in constructable stylesheets.
+
+For these reasons, it is proposed that constructable stylesheets be disabled for the lifetime of the `document` in a post-rehydration context.
+
+### Server-only components
+
+Addressing [Special Case B](#the-happy-path) will involve a solution roughly comparable to [React server components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html). While the bulk of this can be implemented in user-land, at a level of abstraction above LWC, the `hydrateElement` patch implementation will need to support the use-case.
+
+However, we consider support for this case to be out-of-scope for the initial implementation. We will explore further as part of a separate RFC.
+
+### SSR placeholders
+
+Implementation of placeholder components, as described in [Special Case A](#the-happy-path) should require no extra effort, so long as unexpected subtree mismatches (Special Case C) are dealt with properly.

```

Comment:
another alternative that won't generate mismatches is to rely on the fact that on SSR, the `renderedCallback` is not invoked, ex:

```
<template>
    <template if:true={isFirstRender}>
        <p>content placeholder on the server</p>
    </template>
    <template if:false={isFirstRender}>
        <p>rendered only on the client</p>
    </template>
</template>

export default class ClientOnly extends LightningElement {
    isFirstRender = true;
    renderedCallback() {
        this.isFirstRender = false;
    }
}
```